src/node/dynamics_compressor.rs

```diff
@@ -0,0 +1,512 @@
+        let makeup_gain = lin_to_db(full_range_makeup.powf(0.6));
+
+        let mut prev_detector_value = self.prev_detector_value;
+
+        let mut reduction_gain = 0.; // dB
+        let mut reduction_gains = [0.; 128]; // lin
+        let mut detector_values = [0.; 128]; // lin
+
+        for i in 0..RENDER_QUANTUM_SIZE {
+            // pick highest value for this index across all input channels
+            // @tbc - this seems to be what is done in chrome
+            let mut max = f32::MIN;
+
+            for channel in input.channels().iter() {
+                let sample = channel.as_slice()[i];
+                if sample > max {
+                    max = sample;
+                }
+            }
+
+            // pick absolute value and convert to dB domain
+            // var xG in paper
+            let abs = max.abs();
+            let sample_db = lin_to_db(abs);
+
+            // Gain Computer stage
+            // ------------------------------------------------
+            // var yG - eq. 4 in paper
+            // if knee == 0. (hard knee), the `else if` branch is bypassed
+            let sample_attenuated = if sample_db <= threshold - half_knee {
+                sample_db
+            } else if sample_db <= threshold + half_knee {
+                sample_db + (sample_db - threshold + half_knee).powi(2) * knee_partial
+            } else {
+                threshold + (sample_db - threshold) / ratio
+            };
+            // variable xL in paper
+            let sample_attenuation = sample_db - sample_attenuated;
+
+            // Level Detector stage
+            // ------------------------------------------------
+            // Branching peak detector - eq. 16 in paper - var yL
+            // attack branch
+            let detector_value = if sample_attenuation > prev_detector_value {
+                attack_tau * prev_detector_value + (1. - attack_tau) * sample_attenuation
+            // release branch
+            } else {
+                release_tau * prev_detector_value + (1. - release_tau) * sample_attenuation
+            };
+
+            detector_values[i] = detector_value;
+            // cdB = -yL + make up gain
+            reduction_gain = -1. * detector_value + makeup_gain;
+            // convert to lin now, so we just to multiply samples later
+            reduction_gains[i] = db_to_lin(reduction_gain);
+            // update prev_detector_value for next sample
+            prev_detector_value = detector_value;
+        }
+
+        // update prev_detector_value for next block
+        self.prev_detector_value = prev_detector_value;
+        // update reduction shared w/ main thread
+        self.reduction.store(reduction_gain);
+
+        // store input in delay line
+        self.ring_buffer[self.ring_index] = input;
+
+        // apply
+        let read_index = (self.ring_index + 1) % ring_size;
+        let delayed = &self.ring_buffer[read_index];
+
```

Comment:
I think at this point, you could say `*output = delayed.clone()` and then modify these values in place with the `reduction_gains`, no? Would save a few lines

Reply:
Yup true, but I'm not really sure I'm really convinced with this block sized delay, it's way more simple and efficient but it also introduce differences according to the sample rate. 
Maybe we should apply sub quantum logic (without going into sub-sample interpolation) to stick closer to the 6ms delay whatever the sample rate?

Reply:
I'm fine with incremental improvements. Merging this!